storybookjs/storybook
# Core: Improve support for main.ts/preview.ts files

Issue:
It started when I was migrating storybook in my project from 5.2 to 5.3. I had `config.ts` working fine, so when I changed it to `main`, I figured it could as well be `.ts`, however storybook then stopped working. After quite some debugging, I noticed that [at this line](https://github.com/storybookjs/storybook/blob/next/lib/core/src/server/utils/server-require.js#L84), 
`registerCompiler(moduleDescriptor)`, based on the file extension, tries to check if any of the following libs are present under node_modules, traversing from storybook/core folder and going up, until it reaches the user's project folder:
![image](https://user-images.githubusercontent.com/1671563/76357741-8a5df600-6318-11ea-9b41-66b1d505814e.png)
 
If none of those are present, then it won't be able to load the file extension.

## What I did
After talking with @shilman about it, we thought it could be an idea to have that dependency in storybook/core, so the issue is fixed automatically for everyone, rather than having a developer install one of those dependencies in case they would like ts support for those files. So this PR adds that dependency, and with that, any of the storybook configuration files can be used with [typescript](
https://github.com/gulpjs/interpret/blob/master/index.js#L146). The dependency weighs [241 bytes](https://bundlephobia.com/result?p=@babel/register@7.8.3).

## How to test
To test this in the storybook monorepo is a bit complicated, given that many libs have either @babel/register or ts-node as dependencies or have dependencies which have those as peer dependencies, which makes those dependencies available at the top level and it's enough to make the .ts files load without issues. I did test there though, but I had to remove all those dependencies.

 An easier way:
1 - Checkout this [sample repo](https://github.com/yannbf/storybook-with-ts.git)
2 - run `yarn && yarn storybook` and see the storybook working fine.
3 - checkout  the *broken main.ts version*: `git checkout 1e98a7d` or `git checkout HEAD^1`
4 - run `yarn` to uninstall `@babel/register`
5 - run `yarn storybook`and see the issue:
```sh
info @storybook/react v5.3.14
info 
info => Loading static files from: /Users/Dev/storybook-cra-ts/public .
WARN => File /Users/Dev/storybook-cra-ts/.storybook/main.ts is detected
WARN    but impossible to import loader for .ts
info => Loading presets
WARN => File /Users/Dev/storybook-cra-ts/.storybook/main.ts is detected
WARN    but impossible to import loader for .ts
info => Loading presets
```

Alternatively, you can go to the broken version, and rather than installing @babel/register at the project level, install it under `node_modules/@storybook/core`. It will work fine as well.
